dali/kernels/imgproc/convolution/cutlass/kernel/gemm.h

```diff
@@ -152,153 +184,209 @@ struct Gemm {
-                                       tb_offset_A);
+      // We need to start at aligned tile, otherwise tensor ops aren't happy.
+      // Take this into account when calculating the non-zero region
+      // For right side conv-matrix the non-zero regions starts at (n() - left_span, n()),
+      // for the left side it's (m(), m() - left_span)
+      int conv_diag_position = kIsInnerConv ? threadblock_tile_offset.n() * Mma::Shape::kN
+                                          : threadblock_tile_offset.m() * Mma::Shape::kM;
+      constexpr int tile_extent = kIsInnerConv ? Mma::Shape::kN : Mma::Shape::kM;
 
-    typename Mma::IteratorB iterator_B(params.params_B, params.ref_B.data(),
-                                       {problem_size_k, params.problem_size.n()}, thread_idx,
-                                       tb_offset_B);
+      // in this row/column we cover a rectangle starting at (diag - left_span), and ending at
+      // (diag + tile_extent + right_span)
 
-    // Broadcast the warp_id computed by lane 0 to ensure dependent code
-    // is compiled as warp-uniform.
-    int warp_idx = __shfl_sync(0x1f, threadIdx.x / 32, 0);
-    int lane_idx = threadIdx.x % 32;
+      // this handles the border, we don't extend the matrix shapes, where the data starts
+      int k_skipped_offset = max(0, conv_diag_position - left_span);
+      // where we should start on tile boundary
+      k_skipped_offset = (k_skipped_offset / Mma::Shape::kK) * Mma::Shape::kK;
 
-    //
-    // Main loop
-    //
+      cutlass::MatrixCoord tb_offset_A{threadblock_tile_offset.m() * Mma::Shape::kM,
+                                       k_skipped_offset};
 
-    // Construct thread-scoped matrix multiply
-    Mma mma(shared_storage.main_loop, thread_idx, warp_idx, lane_idx);
+      cutlass::MatrixCoord tb_offset_B{k_skipped_offset,
+                                       threadblock_tile_offset.n() * Mma::Shape::kN};
 
-    typename Mma::FragmentC accumulators;
+      // Problem size is a function of threadblock index in the K dimension
+      int problem_size_k =
+          min(params.problem_size.k(), (threadblock_tile_offset.k() + 1) * params.gemm_k_size);
+      int total_gemm_k_iterations = (problem_size_k + Mma::Shape::kK - 1) / Mma::Shape::kK;
 
-    accumulators.clear();
+      // where the data ends
+      int k_end_offset = min(problem_size_k - 1, conv_diag_position + tile_extent + right_span);
+      int k_last_iter = ((k_end_offset + Mma::Shape::kK - 1) / Mma::Shape::kK);
 
-    if (!kSplitKSerial || gemm_k_iterations > 0) {
-      // Compute threadblock-scoped matrix multiply-add
-      mma(gemm_k_iterations, accumulators, iterator_A, iterator_B, accumulators);
-    }
+      // this is how many iterations we need if we start at the offset
+      int start_iteration =
+          (problem_size_k - k_skipped_offset + Mma::Shape::kK - 1) / Mma::Shape::kK;
+      int end_iteration = max(total_gemm_k_iterations - k_last_iter - 1, 0);
```

Comment:
Shouldn't it be
`gemm_k_iterations` and `skip_last_k_iterations`? Now it's very hard to understand, since start_iteration is _larger_ than end_iteration.

Reply:
Renamed, hope it's clearer now.